File: Data-Structures/05_Binary_Search/01_Search_Array.cpp
#include <iostream>
#include <vector>

using std::vector;

int binarySearch(vector<int> arr, int n, int target) {
    int L = 0;
    int R = n;

    while (L <= R) {
        int mid = (L + R) / 2;

        if (target > arr[mid]) {
            L = mid + 1;
        } else if (target < arr[mid]) {
            R = mid - 1;
        } else {
            return mid;
        }
    }

    return -1;
}

int main() {
    vector<int> arr = {1, 3, 3, 4, 5, 6, 7, 8};
    int n = arr.size();
    int target = 7;

    int res = binarySearch(arr, n, target);

    if (res == -1) {
        printf("Target value %d not found", target);
    } else {
        printf("Target value %d found at index %d", target, res);
    }

    return 0;
}

File: DSA-for-Beginners/05_Binary_Search/01_Search_Array.py
def binarySearch(arr, target):
    L, R = 0, len(arr) - 1

    while L <= R:
        mid = (L + R) // 2

        if target > arr[mid]:
            L = mid + 1
        elif target < arr[mid]:
            R = mid - 1
        else:
            return mid
    
    return -1

def main():
    arr = [1, 3, 3, 4, 5, 6, 7, 8]
    target = 7

    res = binarySearch(arr, target)
    if res == -1:
        print(f"Target value {target} not found")
    else:
        print(f"Target value {target} found at index {res}")

if __name__ == "__main__":
    main()